lib/galaxy/datatypes/isa.py

```diff
@@ -0,0 +1,520 @@
+
+    def generate_primary_file(self, dataset=None):
+        """Generate the primary file. It is an HTML file containing description of the composite dataset
+           as well as a list of the composite files that it contains."""
+
+        if dataset:
+            rval = ['<html><head><title>ISA Dataset </title></head><p/>']
+            if hasattr(dataset, "extra_files_path"):
+                rval.append('<div>ISA Dataset composed of the following files:<p/><ul>')
+                for cmp_file in os.listdir(dataset.extra_files_path):
+                    opt_text = ''
+                    rval.append('<li><a href="%s" type="text/plain">%s</a>%s</li>' % (cmp_file, cmp_file, opt_text))
+                rval.append('</ul></div></html>')
+            else:
+                rval.append('<div>ISA Dataset is empty!<p/><ul>')
+            return "\n".join(rval)
+        return "<div>No dataset available</div>"
+
+    # Dataset content needs grooming {{{2
+    ################################################################
+
+    def dataset_content_needs_grooming(self, file_name):
+        """This function is called on an output dataset file after the content is initially generated."""
+        return os.path.basename(file_name) == ISA_ARCHIVE_NAME
+
+    # Groom dataset content {{{2
+    ################################################################
+
+    def groom_dataset_content(self, file_name):
+        """This method is called by Galaxy to extract files contained in a composite data type."""
+        # XXX Is the right place to extract files? Should this step not be a cleaning step instead?
+        # Could extracting be done earlier and composite files declared as files contained inside the archive
+        # instead of the archive itself?
+
+        # extract basename and folder of the current file whose content has to be groomed
+        basename = os.path.basename(file_name)
+        output_path = os.path.dirname(file_name)
+        # extract archive if the file corresponds to the ISA archive
+        if basename == ISA_ARCHIVE_NAME:
+            # perform extraction
+            with open(file_name, 'rb') as stream:
+                self._extract_archive(stream, output_path=output_path)
+            # remove the original archive file
+            os.remove(file_name)
+
+    # Set meta {{{2
+    ################################################################
+
+    def set_meta(self, dataset, overwrite=True, **kwd):
+        """Set meta data information."""
+        super(_Isa, self).set_meta(dataset, **kwd)
+        self._set_dataset_name(dataset)
+        return True
+
+    # Set dataset name {{{2
+    ################################################################
+
+    def _set_dataset_name(self, dataset):
+        investigation = self._get_investigation(dataset)
+        if investigation is not None:
+            dataset.name = investigation.identifier
+        else:
+            dataset.name = 'ISA DATASET'
```

Comment:
Would it be better to create some metadata on this datatype and set this there - I'm a bit worried about overriding the behavior of the upload datatype and the names it creates on a per-datatype basis. I do get this increases the usability of this datatype - but suddenly the upload API isn't going to work the way on would expect if you set a name explicitly or something - and we are constantly debating how to build default names from URL uploads - this would seem to override any decision on that as well. Even if you can make a case for using the investigation if set, it shouldn't be setting the default like this is the investigation is not set right?

Reply:
The `else` part is certainly not useful.
I've noticed that the uploader set the dataset name to the file name (for instance `MTBLS354.zip`), now. I don't remember it was doing that before, but maybe I'm wrong. The filename should be sufficient for the user I guess.